File: src/main/java/ro/fasttrackit/homework/PersonService.java
package ro.fasttrackit.homework;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;

import static java.util.stream.Collectors.*;

public class PersonService {

    private final List<Person> people;

    public PersonService() {
        people = new ArrayList<>();
    }

    public void addPerson(Person person) {
        this.people.add(person);
    }

    public List<Person> getPeople() {
        return new ArrayList<>(people);
    }

    public List<String> getAllNames() {
        return people.stream()
                .map(this::firstAndLastNames)
                .collect(toList());
    }

    public List<Person> getMajorPersons() {
        return people.stream()
                .filter(person -> person.getAge() >= 18)
                .collect(toList());
    }

    public List<Person> personsFromOradea() {
        return people.stream()
                .filter(person -> isFrom(person, "oradea"))
                .collect(toList());
    }

    public List<Person> fromOradeaOrCluj() {
        return people.stream()
                .filter(person -> isFrom(person, "oradea") || isFrom(person, "cluj"))
                .collect(toList());
    }

    public List<String> capitalizedFirstNames() {
        return people.stream()
                .map(person -> person.getFirstName().toUpperCase())
                .collect(toList());
    }

    public List<String> formattedNames() {
        return people.stream()
                .map(this::formatNames)
                .collect(toList());
    }

    public List<Person> withinAgeRange() {
        return people.stream()
                .filter(this::isInAgeRange)
                .collect(toList());
    }

    public Set<String> uniqueFirstNames() {
        return people.stream()
                .map(Person::getFirstName)
                .collect(toSet());
    }

    public List<Person> sortedByFirstName() {
        return people.stream()
                .sorted(Comparator.comparing(Person::getFirstName))
                .collect(toList());
    }

    public List<Person> sortedByLastName() {
        return people.stream()
                .sorted(Comparator.comparing(Person::getLastName))
                .collect(toList());
    }

    public List<Person> completeSort() {
        return people.stream()
                .sorted(Comparator.comparing(Person::getFirstName)
                        .thenComparing(Person::getLastName)
                        .thenComparing(Person::getAge))
                .collect(toList());
    }

    public List<Person> firstNameStartingWithLetter(char letter) {
        return people.stream()
                .filter(person -> firstNameStartsWithLetter(person, letter))
                .collect(toList());
    }

    private String firstAndLastNames(Person person) {
        return person.getFirstName() + " " + person.getLastName();
    }

    private String formatNames(Person person) {
        return person.getFirstName() + " " + person.getLastName().charAt(0) + ".";
    }

    private boolean isInAgeRange(Person person) {
        return person.getAge() > 18 && person.getAge() < 60;
    }

    private boolean isFrom(Person person, String city) {
        return person.getCity().equalsIgnoreCase(city);
    }

    private boolean firstNameStartsWithLetter(Person person, char letter) {
        return person.getFirstName().toLowerCase().charAt(0) == letter || person.getFirstName().toUpperCase().charAt(0) == letter;
    }
}

File: src/main/java/ro/fasttrackit/kotlinhomework/PersonService.kt
package ro.fasttrackit.kotlinhomework

data class Person(val firstName: String, val lastName: String, val age: Int, val city: String) {
    val validator = PersonValidator().verify(firstName, lastName, age, city)
}

class PersonService(private val persons: MutableList<Person> = mutableListOf()) {

    fun addPerson(person: Person) = this.persons.add(person)

    fun getAllPersons(): List<Person> = persons

    fun getAllNames(): List<String> = persons.map { firstAndLastName(it) }

    fun getMajorPersons(): List<Person> = persons.filter { it.age >= 18 }

    fun personsFromOradea(): List<Person> = persons.filter { isFrom(it, "oradea") }

    fun personsFromOradeaAndCluj(): List<Person> = persons.filter { isFrom(it, "oradea") || isFrom(it, "cluj") }

    fun capitalizedFirstNames(): List<String> = persons.map { it.firstName.toUpperCase() }

    fun formattedNames(): List<String> = persons.map { formatName(it) }

    fun personsWithinAgeRange(): List<Person> = persons.filter { it.age in 19..59 }

    fun uniqueFirstNames(): Set<String> = persons.map { it.firstName }.toSet()

    fun sortByFirstName(): List<Person> = persons.sortedBy { it.firstName }

    fun sortByLastName(): List<Person> = persons.sortedBy { it.lastName }

    fun completeSort(): List<Person> = persons.sortedWith(compareBy({ it.firstName }, { it.lastName }, { it.age }))

    fun firstNamesStartingWithLetter(letter: Char): List<Person> = persons.filter { firstNameStartsWithLetter(it, letter) }

    private fun firstAndLastName(person: Person): String = "${person.firstName} ${person.lastName}"

    private fun isFrom(person: Person, city: String) = person.city.toLowerCase() == city.toLowerCase()

    private fun formatName(person: Person): String = "${person.firstName} ${person.lastName[0]}."

    private fun firstNameStartsWithLetter(person: Person, letter: Char): Boolean =
            person.firstName.toLowerCase()[0] == letter || person.firstName.toUpperCase()[0] == letter
}